Models/MuZero_agent_2.py

```diff
@@ -584,88 +589,136 @@ def player_to_play(player_num, action):
-                    search_path[i].append(node[i])
-            
+    
+    def run_batch_mcts(self, roots_cpp, hidden_state_pool):
+        # preparation
+        num = roots_cpp.num
+        # config variables
+        discount = config.DISCOUNT  
+        pb_c_init = config.PB_C_INIT
+        pb_c_base = config.PB_C_BASE
+        hidden_state_index_x = 0
+
+
+        # minimax value storage data structure 
+        min_max_stats_lst = tree.MinMaxStatsList(num)
+        min_max_stats_lst.set_delta(config.MAXIMUM_REWARD*2)  # config.MINIMUM_REWARD *2 
+        horizons = 1  # self.config.lstm_horizon_len, seems to be the number of timesteps predicted in the future 
+        hidden_state_pool = [hidden_state_pool]
+        for _ in range(config.NUM_SIMULATIONS): # go through the tree NUM_SIMULATIONS times 
+            # prepare a result wrapper to transport results between python and c++ parts
+            hidden_states = [] 
+            results = tree.ResultsWrapper(num)
+
+            # evaluation for leaf nodes, traversing across the tree and updating values
+            hidden_state_index_x_lst, hidden_state_index_y_lst, last_action = \
+                tree.batch_traverse(roots_cpp, pb_c_base, pb_c_init, discount, min_max_stats_lst, results)
+            search_lens = results.get_search_len()
+
+            # obtain the states for leaf nodes
+            for ix, iy in zip(hidden_state_index_x_lst, hidden_state_index_y_lst):
+                hidden_states.append(hidden_state_pool[ix][iy])
             # Inside the search tree we use the dynamics function to obtain the next
             # hidden state given an action and the previous hidden state.
-            parent = [search_path[i][-2] for i in range(self.NUM_ALIVE)]
-            hidden_state = np.asarray([parent[i].hidden_state for i in range(self.NUM_ALIVE)])
-            last_action = np.asarray([history[i].last_action() for i in range(self.NUM_ALIVE)])
-
-            network_output = self.network.recurrent_inference(hidden_state, last_action)  # 11.05s
-            for i in range(self.NUM_ALIVE):
-                
-                self.batch_expand_node(node[i], node[i].to_play, network_output)  # 7s
-                
-                # print("value {}".format(network_output["value"]))
-                self.backpropagate(search_path[i], network_output["value"].numpy()[i], min_max_stats[i], i)  # 12.08s
-                
-    def batch_policy(self, observation, prev_action):
-        self.NUM_ALIVE = observation.shape[0]
-        root = [Node(0) for _ in range(self.NUM_ALIVE)]
-        network_output = self.network.initial_inference(observation)  # 2.1 seconds
-        
-        for i in range(self.NUM_ALIVE):
-            self.batch_expand_node(root[i], i, network_output)  # 0.39 seconds
 
-            self.add_exploration_noise(root[i])
-            
-        self.run_batch_mcts(root, prev_action)  # 24.3 s (3 seconds not in that)
+            last_action = np.asarray(last_action)
```

Comment:
These next few lines might be their own method. 636 - 657

Reply:
I dont quite see the benefit of that; it would be a method takes in a lot of inputs and returns a lot of outputs, but doesnt make the code much more legible, and can make debugging harder when trying to track where a variable moves through the program